_drafts/gnu-make-guidelines.md

```diff
@@ -0,0 +1,845 @@
+make -l $CPUCOUNT
+```
+
+## Anatomy of a Makefile
+
+A Makefile contains **rules** used to produce **targets**.
+
+Some basic components of a Makefile are shown below:
+
+```makefile
+# Comments are prefixed with the '#' symbol
+
+# A variable assignment
+FOO = "hello there!"
+
+# A rule creating target "test", with "test.c" as a prerequisite
+test: test.c
+	# The contents of a rule is called the "recipe", and is typically composed
+	# of one or more shell commands.
+	# It must be indented from the target name (historically with tabs, spaces
+	# are permitted)
+
+	# Using the variable "FOO"
+	echo $(FOO)
+
+	# Calling the C compiler using a predefined variable naming the default C
+	# compiler, '$(CC)'
+	$(CC) test.c -o test
+```
+
+Let's take a look at each part of the example above!
+
+## Variables
+
+Variables are used with the syntax `$(FOO)`, where `FOO` is the variable name.
+
+Variable contain purely strings, Make language does not have other data types.
+
+Appending to a variable will add a space and the new content:
+
+```makefile
+FOO = one
+FOO += two
+# FOO is now "one two"
+
+FOO = one
+FOO = $(FOO)two
+# FOO is now "onetwo"
+```
+
+### Variable assignment
+
+In GNU Make syntax, variables are assigned with two "**flavors**":
+
+1. *recursive expansion*: value is set when the variable is consumed
+   ```makefile
+   FOO = 1
+   BAR = foo=$(FOO)
+   FOO = 2
+   $(info BAR=$(BAR))
+   # prints BAR=foo=2
+   ```
+2. *simple expansion*: value is set when Make processes the expression
+   ```makefile
+   FOO = 1
+   BAR := foo=$(FOO)
+   FOO = 2
+   $(info BAR=$(BAR))
+   # prints BAR=foo=1
+   ```
+
+*Note: the `$(info)` function is being used above to print stuff out, and can be
+handy when debugging Makefiles!*
+
+Variables which are not explicitly or implicitly or automatically set will
+evaluate to an empty string.
+
+### Environment variables
+
+Environment variables are carried into the Make execution environment, for
+example:
+
+Makefile contents:
+
+```makefile
+$(info YOLO variable = $(YOLO))
+```
+
+Setting the variable in the shell command when running `make`:
+
+```bash
+$ YOLO="hello there!" make
+YOLO variable = hello there!
+make: *** No targets.  Stop.
+```
+
+*Note: Make prints the "No targets" error because our Makefile had no targets
+listed!*
+
+If you use the `?=` assignment syntax, Make will only assign that value if the
+variable doesn't already have a value:
+
+Makefile:
+
+```makefile
+# default CC to gcc
+CC ?= gcc
+```
+
+We can then override `$(CC)` in that Makefile:
+
+```bash
+$ CC=clang make
+```
+
+Another common pattern is to allow inserting additional flags; in the Makefile,
+append to the variable instead of directly assigning to it, which permits
+passing flags in from the environment:
+
+```makefile
+CFLAGS += -Wall
+```
+
+Then we can add extra flags on our command:
+
+```bash
+CFLAGS='-Werror=conversion -Werror=double-promotion' make
+```
+
+This can be very useful!
+
+### Overriding variables
```

Comment:
Maybe: variables set with `?= or =` will be overridden, ones with `:=` will not be.

Reply:
overriding value will also override simple expansion:
```makefile
FOO := simple_expansion
$(info $(FOO))
```
```bash
❯ make FOO=overridden
overridden
make: *** No targets.  Stop.
```